### DISCUSSION TITLE
Process manager versioning

### DISCUSSION THREAD

**Discussion by slashdotdash** [commanded/commanded#572] [Ideas]
At some point after deployment of a process manager you may realise that you need additional data in its state, available from events already seen/ignored.

Greg Young describes a number of [strategies for versioning process managers](https://leanpub.com/esversioning/read#leanpub-auto-versioning-process-managers):

- Replace the existing process manager with an entirely new one.
- Upcast its state.

Another approach is to event source process manager state allowing it to be rebuilt whenever required (#127). This can be combined with state snapshotting as a performance optimisation.

To support upcasting of state, Commanded could: replay a process manager's events to allow each instance's state to be rebuilt (and *not* dispatch commands during replay); or provide an `upgrade` function that is called to mutate the process manager's state to a new version; or event source its state.

**Comment by dvic** [comment]
Description mentions #127, but I think you mean #137 here?
